File: testing/workflows/scripts/extract_docx.py
#!/usr/bin/env python3
"""Extract translatable text units from a DOCX file to JSON."""
import json
import re
import sys
from pathlib import Path
from docx import Document

CODE_PATTERN = re.compile(r"^\s*\d{3}-\d{5}[A-Z]?\s*$|^\s*\d{6}\s*$")

def extract_units(docx_path: str) -> list:
    doc = Document(docx_path)
    units = []
    
    # Body paragraphs
    for pi, p in enumerate(doc.paragraphs):
        style = p.style.name if p.style else ""
        for ri, run in enumerate(p.runs):
            text = run.text
            if text and text.strip() and not CODE_PATTERN.match(text.strip()):
                units.append({
                    "id": f"body:p{pi}:r{ri}",
                    "source": text,
                    "style": style,
                    "where": "body"
                })
    
    # Headers/Footers
    for si, section in enumerate(doc.sections):
        for pi, p in enumerate(section.header.paragraphs):
            for ri, run in enumerate(p.runs):
                text = run.text
                if text and text.strip():
                    units.append({
                        "id": f"header:s{si}:p{pi}:r{ri}",
                        "source": text,
                        "style": p.style.name if p.style else "",
                        "where": "header"
                    })
        
        for pi, p in enumerate(section.footer.paragraphs):
            for ri, run in enumerate(p.runs):
                text = run.text
                if text and text.strip():
                    units.append({
                        "id": f"footer:s{si}:p{pi}:r{ri}",
                        "source": text,
                        "style": p.style.name if p.style else "",
                        "where": "footer"
                    })
    
    return units

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python extract_docx.py <input.docx> <output.json>")
        sys.exit(1)
    
    units = extract_units(sys.argv[1])
    Path(sys.argv[2]).write_text(json.dumps(units, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Extracted {len(units)} units -> {sys.argv[2]}")

File: testing/workflows/scripts/extract_docx.mjs
#!/usr/bin/env node
/**
 * Extract translatable text units from a DOCX file to JSON.
 * Uses mammoth for DOCX parsing.
 */
import mammoth from 'mammoth';
import fs from 'fs/promises';
import path from 'path';

const CODE_PATTERN = /^\s*\d{3}-\d{5}[A-Z]?\s*$|^\s*\d{6}\s*$/;

/**
 * Extract translation units from DOCX
 * mammoth extracts as HTML, we parse paragraphs from that
 */
async function extractUnits(docxPath) {
  const buffer = await fs.readFile(docxPath);
  
  // Get raw text with paragraph markers
  const result = await mammoth.extractRawText({ buffer });
  const text = result.value;
  
  // Also get HTML for structure detection
  const htmlResult = await mammoth.convertToHtml({ buffer });
  const html = htmlResult.value;
  
  const units = [];
  
  // Split by paragraphs (double newlines or single newlines)
  const paragraphs = text.split(/\n+/).filter(p => p.trim());
  
  paragraphs.forEach((para, pi) => {
    const trimmed = para.trim();
    
    // Skip empty or code-only patterns
    if (!trimmed || CODE_PATTERN.test(trimmed)) {
      return;
    }
    
    // Detect style from HTML (basic heuristic)
    let style = 'body';
    if (html.includes(`<h1>${trimmed.substring(0, 20)}`)) style = 'heading1';
    else if (html.includes(`<h2>${trimmed.substring(0, 20)}`)) style = 'heading2';
    else if (html.includes(`<strong>${trimmed.substring(0, 20)}`)) style = 'bold';
    
    units.push({
      id: `body:p${pi}`,
      source: trimmed,
      style: style,
      where: 'body'
    });
  });
  
  return units;
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 2) {
    console.log('Usage: node extract_docx.mjs <input.docx> <output.json>');
    process.exit(1);
  }
  
  const [inputPath, outputPath] = args;
  
  try {
    const units = await extractUnits(inputPath);
    await fs.writeFile(outputPath, JSON.stringify(units, null, 2), 'utf-8');
    console.log(`Extracted ${units.length} units -> ${outputPath}`);
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

main();
